### PR TITLE
303 Added api to get block span list

### LINKED ISSUE #303: Create a API which expose the list of blockspans
Create a function that takes 3 arguments

Function signature:

blockspans(end, gap, number)or could be (start, gap, number)

Arguments: 

- Start block height
- Span length
- Number of spans

### PR DESCRIPTION
**Pull request by Natsvora**
https://github.com/dambaev/op-energy/issues/303 

Added API which returns block span list. Also updated the nbdr calculation to get block spans based on newly created function

### CODE REVIEW

File: backend/src/oe/api/op-energy.service.ts [hunk 1]
```diff
@@ -444,6 +444,23 @@ export class OpEnergyApiService {
     return await bitcoinApi.$getBlock(hash.value);
   }
 
+  public async $getBlockSpanList(UUID: string, startBlockHeight: number, span: number, numberOfSpan: number): Promise<BlockSpan[]> {
```
Comment (dambaev) [inline backend/src/oe/api/op-energy.service.ts]: why is this function async? I see no particular reason for it to be async...

File: backend/src/oe/api/op-statistics.service.ts [hunk 2]
```diff
@@ -19,36 +21,44 @@ export class OpStatisticService {
...
+              }
+            )
           )
         );
 
         const blockHeadersList =
           await opBlockHeaderService.$getBlockHeadersByHeights(
             'nbdr',
-            confirmedBlocks
+            Array.from(new Set(confirmedBlocks))
```
Comment (dambaev) [inline backend/src/oe/api/op-statistics.service.ts:56]: why do you convert `ConfirmedBlockHeight[]` into `Set<ConfirmedBlockHeight>` and back to `ConfirmedBlockHeight[]`? to remove duplicates?

File: backend/src/oe/api/routes.ts [hunk 3]
```diff
@@ -212,6 +212,20 @@ class OpEnergyRoutes {
     }
     logger.info( `${UUID}: PROFILE: end: $getBlockByHash`);
   }
+
+  private async $getBlockSpanList(req: Request, res: Response) {
+    const UUID = await opEnergyApiService.$generateRandomHash();
+    try {
+      logger.info( `${UUID}: PROFILE: start: $getBlockSpanList`);
+      const { startBlockHeight, span, numberOfSpan } = req.params;
```
Comment (dambaev) [inline backend/src/oe/api/routes.ts:221]: I am worried about an absence of input validation here: either of `startBlockHeight, span, numberOfSpan` can have string value and such variables will have `NaN` values. If any of those variables will be negative, then they will be passed as negative numbers further to opEnergyApiService.$getBlockSpanList() call.

**Review by dambaev — CHANGES_REQUESTED** [review]

File: backend/src/oe/api/op-energy.service.ts (hunk 1 above)
Comment (Natsvora) [reply backend/src/oe/api/op-energy.service.ts]: Fixed

File: backend/src/oe/api/op-statistics.service.ts (hunk 2 above)
Comment (Natsvora) [reply backend/src/oe/api/op-statistics.service.ts:56]: Yes, correct

File: backend/src/oe/api/routes.ts (hunk 3 above)
Comment (Natsvora) [reply backend/src/oe/api/routes.ts:221]: done

File: backend/src/oe/api/routes.ts [hunk 4]
```diff
@@ -212,6 +213,29 @@ class OpEnergyRoutes {
...
+
+      if (
+        !isValidPositiveNumber(startBlockHeight) ||
+        !isValidPositiveNumber(span) ||
+        !isValidPositiveNumber(numberOfSpan)
```
Comment (dambaev) [inline backend/src/oe/api/routes.ts]: I agree, that `numberOfSpan` and `span` should be positive, but `startBlockHeight` should be Natural, as we can ask to build spans starting from 0

**Review by dambaev — CHANGES_REQUESTED** [review]

File: backend/src/oe/api/routes.ts (hunk 4 above)
Comment (Natsvora) [reply backend/src/oe/api/routes.ts]: Done

**Review by dambaev — APPROVED** [review]

**Pull request merged**